design.md

```diff
@@ -0,0 +1,109 @@
+```shell
+> teleport --cacert="/path/to/ca.cert" \
+           --key="/path/to/client.key" \ 
+           --cert="/path/to/client.cert" \
+           --addr="server address" 
+           COMMAND [OPTIONS] [ARG...]
+ ```
+`[Options]` is the set of flags or options supported by the cli and `[ARG...]` is the command name and arguments of the command the user wishes to run on the remote system. Example flow with server address `localhost:3000` and `ping` executable below. 
+### start 
+`start` starts a process on the server. An id identify the process is returned to be used with the other commands. Specify the `-l` flag to listen for process output after starting the process. This will return a job id of `job1` for example.
+```shell
+> teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" \
+           start -l ping google.com
+```
+### status
+`status` returns the status of a job. Specify the job id returned from the `start` command to run `status`.
+```shell
+teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" status job1
+```
+### output
+`output` displays the output from a process. Specify the job id returned from the `start` command to run `output`.
+```shell
+teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" output job1
+```
+### stop
+`stop` stops a process running on the server. Specify the job id returned from the `start` command to run `stop`.
+```shell
+teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" stop job1
+```
+The full set of commands will be provided through the help message which can be invoked with:
+```shell
+> teleport --help
+```
+
+## Authentication/Authorization
+Client and server communication must be encrypted and secure. mTLS authentication is used to achieve this. A tool will be provided to generate certificates for the client and server using the same CA. The client certificates are each generated with a unique common name so the server can differentiate between different clients.
+### TLS minimum version
+TLS 1.3 is the minimum supported version
+### Cipher suites
+All TLS 1.3 cipher suites are supported
+* TLS_AES_256_GCM_SHA384
+* TLS_CHACHA20_POLY1305_SHA256
+* TLS_AES_128_GCM_SHA256
+* TLS_AES_128_CCM_8_SHA256
+* TLS_AES_128_CCM_SHA256
+### Key generation algorithm
+Keys are generated using 4096 bit RSA
+
+### Authorization
+When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `UNAUTHOURIZED` error.
```

Comment:
nit: error codes can be used by attackers to figure out whether a job ID exists or not.
always prefer `NOT_FOUND`-type errors to `UNAUTHORIZED` when checking permissions

```suggestion
When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `NOT_FOUND` error.
```

Reply:
ok thanks.